### DISCUSSION TITLE
Access modifiers

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#725] [Language design]
I'm opening the discussion about the design of Val's access modifiers. In my experience, access modifiers is a common topic of disagreement so I'm ready for very strong opinions ;)

## Motivation

Access modifiers are an important component of API design. However, traditional approaches tend to be not very flexible and particularly not well-suited to extensions and post-hoc conformance declarations. Even Swift's design has some wrinkles.

Swift has [5 access modifiers](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/accesscontrol/) (from least to most restrictive):

- `open` (*only relevant to Swift's class system*)
- `public`
- `internal` (default)
- `fileprivate`
- `private`

Despite these 5 modifiers, it is not possible to declare an extension that is able to access private members _and_ expose new public entities. Since Swift uses extensions to declare post-hoc conformance, private details of a type must sometimes be exposed more than necessary just to support a conditional conformance declaration (unless the extension is in the same file).

Further, it is also not possible to allow a companion type to access private details of another type. Although this limitation can be mitigated using `fileprivate` and/or some clever nesting, it remains a hurdle in general.

## Proposed approach

We define the following access modifiers:

- `public`
- `internal`
- `scoped(to: S1, ..., Sn)`
- `private` (default)

_Note: `internal` is an alias for `scoped(to: M)` where `M` is the name of the enclosing module._

`public` means exposed outside of the enclosing scope. A `public` declaration is part of a module's API if and only if all its enclosing scopes are public. For example, in the file below, `A` and `A.f` are part of the module's API whereas `A.g`, `B` and `B.f` aren't.

```swift
public namespace A {
  public fun f() {}
  fun g() {}
}
namespace B {
  public fun f() {}
}
```

`scoped(to: S1, ..., S2)` can be used to expose a declaration nested in a scope without making it as visible as that scope. For example, in the file below, `A.g` is visible to the whole module but, unlike `A.f`, it is not part of that module's API.

```swift
public namespace A {
  public fun f() {}
  internal fun g() {}
}
```

`scoped(to: S1, ..., S2)` can also be used to confer access to a companion scope. For example, in the file below, `A.f` is only visible within `A`, `B`, and `C.X`.

```swift
public namespace A {
  scoped(to: B, C.X) fun f()
}
namespace B {}
namespace C {
  public namespace X {}
}
```

Note that a declaration that is `scoped(to: S)` is also visible in the extensions of `S`.

`private` means only visible within the enclosing scope and is the default access level. One exception is made for extension and conformance declarations in the same module, which are allowed to access the private members of the type they extend.

*Note: This exception is more permissive than Swift, where it only applies to extensions in the same file.*

`private` applied to top-level declarations of a file make them visible only to that file. Only `public` and `internal` top-level declarations are visible to the whole module.

## Rationale

I originally thought we could get away with only one access modifier: `public`. A declaration would be private by default and `public` would simply make it visible to enclosing scope. If this scope is also `public`, then the nested declaration is also visible to its enclosing scope, etc. For example:

```swift
// `Foo` is visible in the containing file and isn't part of the module's API.
namespace Foo {
  // `Foo.f` is only visible in the containing file.
  public fun f() {}
  
  // `Foo.A` is only visible within `Foo`.
  type A {
    // `Foo.A.f` is only visible within `Foo`.
    public static fun f() {}
  }
}

// `Bar` is part of the module's API.
public namespace Bar {
  // `Bar.f` is visible part of the module's API.
  public fun f() {}

  // `Bar.g` is only visible within `Bar`.
  fun g() {}
}
```

This approach gets you surprisingly far, but it has a couple of annoying limitations. First, note that members of declarations that are part of a module API can't have members that are only visible within the module. Second, it is not possible to access private declarations in extensions.

For example:

```swift
public type A {
  // `A.f` is part of the module's public API.
  public fun f() {}
  // `A.g` is only visible within `A`.
  fun g() {}
}

extension A {
  fun h() {
    // `A.g` is not visible here.
  }
}
```

@dabrahams suggested a very nice feature that possibly addresses both issues.

> if our access control ever turns out to be insufficient: we could add something like `scoped(to: SomeScopeName)`

This modifier is reminiscent of C++'s `friend` feature. I think that if we combine it with namespaces, then we can work around the two issues I mentioned.

## Alternative considered

Rather than allowing all extension and conformance declarations access to the private declarations of the extended entity, we could add a `protected` modifier that confers and make `private` stricter. However, the benefits of this feature is questionable.

**Comment by Nobody1707** [comment]
I've been a long proponent of the idea that Swift changing the definition of private was a mistake. I am not convinced that anything more granular than public, internal (the default), and private (i.e. file private) are needed. Especially in a language with actual namespaces.

**Reply by kyouko-taiga** [reply]
It is not exactly clear to me what Swift "changed" about `private`. AFAICT, the way Swift's `private` works is pretty consistent with mainstream programming languages, at least at type scope. Besides, there isn't a canonical definition of what access modifiers are supposed to mean.

I strongly believe that choosing internal as a default was an odd choice in a language that otherwise consistently favors the most restrictive options. I'd like to avoid that in Val.

**Reply by Nobody1707** [reply]
When Swift first got access control in Swift 2, private meant the same thing as fileprivate does now. The renamed private to fileprivate when they added the current version.

**Reply by kyouko-taiga** [reply]
Thanks for the clarification. I would tend to agree with you then. I'd like to add that I do not feel very strongly about access modifiers in general (I was originally a Python developer if anyone can believe it) so I am all for keeping things simple.

**Comment by dabrahams** [comment]
1. I think the behavior of Swift's own access modifiers was updated at some point and that's the change the OP is referring to.
2. Choosing internal as a default is aimed at allowing people to use Swift without worrying about access control or paying for it syntactically (i.e. like Python).  I'm not saying it is necessarily the best design, but I understand the choice.  It's all about tradeoffs.

**Reply by kyouko-taiga** [reply]
Though I understand the goal, I'm not convinced using `internal` to avoid syntactic costs actually works. In Val sources, for example, most definitions need to be `public`. So we pay the syntax tax anyway. Maybe things work better for people who can fit all their code in a single module.

In any case, I think a better way to avoid the tax is something like `@implicit <access-level>`.

**Comment by kyouko-taiga** [comment]
After discussing with @dabrahams, we think we could start with only `public`, `internal`, and `private`. These modifiers would behave as described above with one additional exception: any extension or conformance defined in the same module would have access to private members of the extended type. For example:

```swift
// In A.val
public type A {
  var raw_value: Int
  public memberwise init
}

fun foo(a: A) -> Int {
  a.raw_value.copy() // error: `A.raw_value` is private
}

// In B.val
extension A {
  public fun bar() -> Int {
    raw_value.copy() // OK
  }
}
```
Assuming `A.val` and `B.val` are in the same module, `A.bar` would type check.

**Comment by sisoje** [comment]
Reviving this three years on, because the three-modifier resolution is calcifying toward 1.0 and the thread stopped one step short of its own conclusion.

> "In Val sources, for example, most definitions need to be `public`. So we pay the syntax tax anyway."

That's because per-declaration modifiers are the wrong unit. The alternative isn't fewer modifiers — it's an export list, ML-style: one interface file per module (OCaml's `.mli`, Elm's `exposing`), since a Hylo module spans many files. The opening examples, with zero keywords in the bodies:

    // M.interface — the module's entire API, one file
    exposing (
      A.f,       // A.g, B, B.f: not exported, hence module-private
      A,         // opaque: the type is API, raw_value is not
      A.init,
      A.bar,
    )

    namespace A {
      fun f() {}
      fun g() {}
    }

    type A {
      var raw_value: Int
      memberwise init
    }

    // other file, same module
    extension A {
      fun bar() -> Int { raw_value.copy() }  // OK: same module, no rule needed
    }

The extension case stops being an "exception" — same-module code sees everything because visibility *is* module membership, which is exactly what the resolution here already conceded. Outside the module, `raw_value` is invisible because it isn't in the list. Both opening problems become unrepresentable, the common case costs one line instead of a keyword per declaration, and representation hiding falls out as opaque exports.

The ergonomics compound on the reading side: with sprinkled modifiers, a module's API exists nowhere — you reconstruct it by scrolling every file, and an API review is a keyword hunt across the codebase. The interface file *is* the API: one place to read it, one file to diff when it changes.

`scoped(to:)` is the only real casualty — but that's C++ `friend` by another name. And C++ programmers won't miss anything: they already maintain a public surface in a separate file. The interface file keeps the habit and makes the compiler enforce it.

**`private`'s deeper job — guarding invariants against mutation through aliases — needs no keyword in a language with no aliases.** @dabrahams' law of exclusivity already killed those.

A language bold enough to delete classes is bold enough to delete `public`.

**Reply by dabrahams** [reply]
Seems private does nothing to solve or even mitigate against aliasing, or we could program in Java with no spooky action. 

There's a new scheme that @kyouko-taiga implemented in hylo-new. Please check that out and if you still think we're missing something raise an issue or discussion there. Thanks!

**Reply by tothambrus11** [reply]
Hey, thanks for the comment! See the newest scheme here: https://github.com/hylo-lang/hylo-new/issues/55 By default, everything is fileprivate, and you can explicitly expose declarations if needed. I would be strongly against interface files, after 4 years of working with header files in C++. Keeping up the consistency between the interface file and the implementation is really annoying. IDEs help, but are still inflexible for many workflows. Instead of cutting and pasting a function from one module to another, now you need to move the interface definition as well. Having one file to diff is not any better than having multiple files to diff. Git diff is free.

(For the future: please avoid using AI to phrase your posts in the Hylo community, it helps us connect and understand each other much better.)

**Reply by sisoje** [reply]
> Hey, thanks for the comment! See the newest scheme here: [hylo-lang/hylo-new#55](https://github.com/hylo-lang/hylo-new/issues/55) By default, everything is fileprivate, and you can explicitly expose declarations if needed. I would be strongly against interface files, after 4 years of working with header files in C++. Keeping up the consistency between the interface file and the implementation is really annoying. IDEs help, but are still inflexible for many workflows. Instead of cutting and pasting a function from one module to another, now you need to move the interface definition as well. Having one file to diff is not any better than having multiple files to diff. Git diff is free.
> 
> (For the future: please avoid using AI to phrase your posts in the Hylo community, it helps us connect and understand each other much better.)

Hello, thanks, and sorry for using AI tools - I am not a native English speaker, so I used it mainly for style checking and to make it politically correct :)

I am mostly interested in nice language ergonomics, and what frustrates me in Swift and a lot of supposedly "data-oriented" languages is that they require too much boilerplate of access identifiers for no benefit. I feel its a habit from OOP times.

For example Swift does not generate a public memberwise init for public structs - so you can make a struct public, all of its properties public, and then some bozo forgets to write the public init by hand and it's a total mess. Apple releases such code in production frameworks, which makes it impossible for me to construct pure data in tests.

I will follow up on the new thread, thanks again.

**Reply by tothambrus11** [reply]
Yeah, I've also been bitten by the accidental lack of public constructors in some libraries. That's usually a sign that they haven't tested the library well enough, or they had `@testable import` enabled for more files than necessary. At least in Hylo it's always required to write the memberwise initializer, and changing it to public is just the matter of adding one word:

```hylo
struct Point {
  var x: Float32
  var y: Float64

  memberwise init // add public if needed
}

let p = Point(x: 3.14, y: 0)
```
We may add some shorthand syntax that lets us define record types trivialy (like Java's `record`), but I wouldn't rush committing to more sugar for now.

**Reply by dabrahams** [reply]
> changing it to public is just the matter of adding one word

It probably won't remain so for resilient structs, for the same reasons that Swift doesn't do it.

**Reply by tothambrus11** [reply]
I don't see what you mean. Do you want to ban `public memberwise init`? 

I want to be able to replicate the behavior of a memberwise initializer using a non-sugared version, so that we can reorder the fields without breaking the constructor: the new source with reordered members should expand to an explicit memberwise constructor. See some ideas here: https://github.com/orgs/hylo-lang/discussions/1890

**Reply by dabrahams** [reply]
No. All I mean is that a memberwise init can't resiliently remain memberwise when the members change.
